File: native/ZiNotes/app/src/main/java/com/example/zinotes/data/DataSource.kt
package com.example.zinotes.data

import com.example.zinotes.model.Hanzi


object DataSource {
    var hanziList = mutableListOf<Hanzi>(
        Hanzi(
            "1","zi", listOf(4, 0), 39, 6, 1,
            listOf("character", "letter", "word")
        ),
        Hanzi(
            "2","cha", listOf(2), 140, 9, 1,
            listOf("tea")
        )
    )

    fun findHanziById(id: String): Hanzi? {
        return hanziList.find { it.id == id }
    }

    fun addHanzi(hanzi: Hanzi) {
        hanziList.add(hanzi)
    }

    fun updateHanzi(hanzi: Hanzi) {
        val index = hanziList.indexOfFirst { it.id == hanzi.id }
        if (index != -1) {
            hanziList[index] = hanzi
        }
        else
            addHanzi(hanzi)
    }

    fun deleteHanziById(hanziId: String?) {
        if (hanziId == null) return
        hanziList.removeAll { it.id == hanziId }
    }
}

File: non-native/zinotes_flutter/lib/data/datasource.dart
import '../model/hanzi.dart';

class HanziRepository {
  final List<Hanzi> _hanziList = [
    Hanzi(
        id: "1", pinyin: "zi", tones: [4, 0], radicalNumber: 39, strokeCount: 6,
        hskLevel: 1, definitions: ["character", "letter", "word"]
    ),
    Hanzi(
        id: "2", pinyin:"cha", tones: [2], radicalNumber: 140, strokeCount: 9,
        hskLevel: 1, definitions: ["tea"]
    )
  ];

  List<Hanzi> getHanziList() {
    return _hanziList;
  }

  void addHanzi(Hanzi hanzi) {
    _hanziList.add(hanzi);
  }

  void removeHanzi(Hanzi hanzi) {
    int index = _hanziList.indexOf(hanzi);
    if (index != -1) {
      _hanziList.remove(hanzi);
    }
  }

  void updateHanzi(Hanzi hanzi) {
    int index = _hanziList.indexWhere((item) => item.id == hanzi.id);
    if (index != -1) {
      _hanziList[index] = hanzi;
    }
    else {
      _hanziList.add(hanzi);
    }
  }

  Hanzi? getHanziById(String id) {
    for (Hanzi hanzi in _hanziList) {
      if (hanzi.id == id) {
        return hanzi;
      }
    }
    return null;
  }

}